src/libsyntax/ast.rs

```diff
@@ -1746,13 +1730,21 @@ impl StructFieldKind {
     }
 }
 
+#[derive(Clone, Copy, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
+pub enum VariantKind {
+    Dict,
+    Tuple,
+    Unit,
+}
+
 #[derive(Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash, Debug)]
 pub struct StructDef {
     /// Fields, not including ctor
     pub fields: Vec<StructField>,
     /// ID of the constructor. This is only used for tuple- or enum-like
     /// structs.
-    pub ctor_id: Option<NodeId>,
+    pub id: NodeId,
```

Comment:
Nice, I so almost did this. I think what I am proposing is:

What                |Kind    |NodeId in Variant   |NodeId in VariantKind
-------------------------------------------------------------------------
Variant in Enum     |Unit    |Used                |Used
Variant in Enum     |Tuple   |Used                |Used
Variant in Enum     |Dict    |Used                |N/A
Struct              |Unit    |N/A                 |Used
Struct              |Tuple   |N/A                 |Used
Struct              |Dict    |N/A                 |N/A

But I think we disagree about where a ctor id is used - I thought that Tuple and Unit variants use their ctor id - tuple variants are certainly valid functions in Rust. In the first table there is also duplication between the node id in StructDef and the NodeId in `Item` in the non-variant struct cases (iiuc).